Implemented onInit, onUpdate, reset, and pause

reedperkins opened the pull request:
## What?

I implemented  `app.onInit` and  `app.onUpdate` and added two additional functions (along with their implementations) as part of the task requirements: `app.reset` and `app.pause`. Additionally, I created auxiliary functions and data structures to hold game state and perform business logic such as collision detection.

- [x] Make the game cover the whole browser
- [x]  Add keyboard functionalities for both players, W and S for one player, Up and Down for another player.
- [x]  Add players into canvas, make sure we can move players with keyboard.
- [x]  Add ball to the game, pressing SPACE button should start and pause game, make sure ball bounces from boundries.
- [x] When player scores, show scoring with console.log. And reset game.

Additional features include:

- "Game Paused" displayed in upper right corner when game is paused
- User can resize the window and the paddles will be adjusted to compensate
- Scoreboard is displayed

## Why?

These functions are needed to implement the behavior described in the README.

## How?

State is contained in a single object. Entities (e.g. the ball and paddles) are modeled using Javascript classes. A single transition function is provided that updates the current state of the game based on the coordinates of each entity as well as the current keyboard input.

Collision detection is performed by checking if the coordinates of the ball are equal to the coordinates of a paddle or wall. This style of collision detection is rudimentary. Bugs can be more easily seen when the ball size is enlarged to size `200` and the ball speed is increased -- there is a chance that, because of the interaction between game speed and size of the ball in relation to the paddle, the hit box of the paddle will be missed. This doesn't happen often with a small ball and the current implementation is passable. This could be a source for future work, however. 

## Testing

The main source of testing would be the transition function. My initial thought was to make this function pure by copying the entire game state within the transition function and returning a new object. However, because this game requires an efficient update step, I was worried that this would impact performance. Therefore, the game state is global and mutable which allows for efficient updates, but makes it arguably harder to test algorithmically. Instead, manual testing will need to be performed (see video link below).

Video Link: [https://share.vidyard.com/watch/ExowRZrNHtTZw7rYZxCBB9?](https://share.vidyard.com/watch/ExowRZrNHtTZw7rYZxCBB9?)

reedperkins committed:
Implemented reset

reedperkins committed:
Implemented pause

reedperkins committed:
Implemented onInit

reedperkins committed:
Implemented onUpdate

reedperkins committed:
Adding gitignore and package-lock.json

reedperkins committed:
Added pause overlay and styles

reedperkins committed:
Fixed resize issue

reedperkins committed:
Converted tabs to spaces

demiculus commented:
@reedperkins well done, can you do a few edits to this?

1. Game is not fun, can you make it fun? Feel free to add mechanics, change numbers, do whatever you think would make it more fun
2. Can you clean up the code so that there is no function over 10 lines, while at it rename things well so you don't have to add comments
3. Rename confusing variables to be more explicit.

reedperkins committed:
Added controls section

reedperkins committed:
Added Pong aesthetic

reedperkins committed:
Refactored functions to be (around) 10 lines max

reedperkins commented:
@demiculus Here's what I did
> Game is not fun, can you make it fun? Feel free to add mechanics, change numbers, do whatever you think would make it more fun

To make the game more fun, I added the ability for the use to enable bots to play the game (see video below). In addition, I modified the look of the game to have more of the classic Pong aesthetic! This included the following changes:
1. I changed the font and made the scoreboard layout more sparse (the full message is still printed to the console)
2. I used a black and white color palette (classic Pong style)
3. I added an algorithmically generated center line down the middle
4. Ensured that resizing the window still works as expected
5. Made the ball round

> Can you clean up the code so that there is no function over 10 lines, while at it rename things well so you don't have to add comments

I refactored the code so that (almost all) functions were less than 10 lines

> Rename confusing variables to be more explicit.

I removed the "multiplier" suffix from the variable names and in other cases made them more clear.

Here is a video of the new game in action!

https://share.vidyard.com/watch/np4iQ8QzLbsrFvWvoLDAnc?

(The video is a little laggy, but I promise the resizes are smooth :) )
